TELCODOCS-416: CNF-3205 Telco friendly workload specific PerformanceProfile API WIP (#46293)
openshift/openshift-docs: modules/cnf-configuring-workload-hints.adoc

// Module included in the following assemblies:
//
// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc

:_content-type: CONCEPT
[id="cnf-configuring-workload-hints_{context}"]
= Configuring workload hints

Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:

* Low latency
* Real-time capability
* Efficient use of power
* High throughput

In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload class and better able to meet the demands of a workload class. The cluster admin can now specify into which use case a workload falls. The Node Tuning Operator uses hints fine tune the performance settings.

> The environment in which an application is operating influences the choice of hints. In settings such as a core network, only minimal default power tuning is supported using CPU partitioning. In  distributed or Edge settings, latency is a higher priority, but some measures are still taken to optimize power consumption. For far edge settings, at a plane where where software-based radio is transmitted, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.

Reviewer: > only minimal default power tuning is supported using CPU partitioning
@MarSik  can we state that this is the most common case ?

Reviewer 2: The paragraph needs rewrite, it is jumping from one undefined term to another, mixing concepts etc..

I propose something like this (I edited out all the Telco specific terms that were not defined anywhere):

```suggestion
The environment in which an application is operating influences its behavior. For a typical datacenter with no strict latency requirements, only minimal default tuning is needed that enables CPU partitioning for the few and far between high performance workload pods.
Then there are data centers and workloads where latency is a higher priority, but some measures are still taken to optimize power consumption.
The most complicated cases are clusters close to latency sensitive pieces of equipment like manufacturing machinery, software defined radios etc. This last class of deployment is also known as Far edge. For far edge deployments, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.
```

Author: Thanks for the re-write of my first draft. I made some changes as required by our Style Guide.

> The environment in which an application is operating influences the choice of hints. In settings such as a core network, only minimal default power tuning is supported using CPU partitioning. In  distributed or Edge settings, latency is a higher priority, but some measures are still taken to optimize power consumption. For far edge settings, at a plane where where software-based radio is transmitted, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.

Reviewer: > where where
@tmulquee  remove one `where`

Author: Done.

// Module included in the following assemblies:
//
// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc

:_content-type: CONCEPT
[id="cnf-configuring-workload-hints_{context}"]
= Configuring workload hints

Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:

* Low latency
* Real-time capability
* Efficient use of power

> In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload class and better able to meet the demands of a workload class. The cluster admin can now specify into which use case a workload falls. The Node Tuning Operator uses hints fine tune the performance settings.

Reviewer 2: ```suggestion
In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case his workload falls. The Node Tuning Operator uses hints to fine tune the performance settings for it.
```

Author: Done with minor style guide changes - such as "his" -> "that".

The environment in which an application is operating influences the choice of hints. In settings such as a core network, only minimal default power tuning is supported using CPU partitioning. In  distributed or Edge settings, latency is a higher priority, but some measures are still taken to optimize power consumption. For far edge settings, at a plane where where software-based radio is transmitted, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.

[cols="1,1,1,1",options="header"]
|===
    | Hint | Application | Environment | Description

>     | Default

Reviewer 2: This column needs a header pointing to the Performance profile creator power mode

Author: I suggest preceding the table with a Note: "The following workload hints can be configured manually or created using the Performance Profile Creator. See "Creating a performance profile" for further information."

    a|[source,terminal]
----
workloadHints:
highPowerConsumption: false
realtime: false
----

>     | Core network operations

Reviewer 2: ```suggestion
    | High throughput cluster without latency requirements
```

    | Performance achieved through CPU partitioning only.

    | Low-latency
    a|[source,terminal]
----
workloadHints:
highPowerConsumption: false
realtime: true
----

>     | Edge applications

Reviewer 2: Workloads with modest latency requirements, regional datacenters. I would stay away from simple "Edge", because it is ambiguous for non-Telco folks.

    | Both energy savings and low-latency are desirable: compromise between power management, latency and throughput.

    | Ultra-low-latency
    a|[source,terminal]
----
workloadHints:
highPowerConsumption: true
realtime: true
----

>     | Far edge settings

Reviewer 2: ```suggestion
    | Far edge clusters, latency critical workloads
```

Author: Done.

>     | No power management, minimal latency, and maximum determinism.

Reviewer 2: ```suggestion
    | Optimized for absolute minimal latency and maximum determinism at the cost of increased power consumption.
```

Author: Done.

|===

.Procedure

> . Create a profile appropriate for the environment's hardware and topology. In this example, we tune for ultra-low-latency.

Reviewer 2: ```suggestion
. Create a PerformanceProfile appropriate for the environment's hardware and topology as described in chapters above. Then tweak the profile to match the expected workload. In this example, we tune for the lowest possible latency.
```

Author: Done.

> . Add the `highPowerConsumption` and `realtime` parameters for ultra-low-latency workload hints:

Reviewer 2: ```suggestion
. Add the `highPowerConsumption` and `realtime` workload hints and set them both to true according to the table above.
```

Author: Done.

+
[source,yaml]
----
    apiVersion: performance.openshift.io/v2
    kind: PerformanceProfile
    metadata:
      name: workload-hints
    spec:
      ...
      workloadHints:
        highPowerConsumption: true <1>
        realtime: true <2>
----

> <1> If `highPowerConsumption` is true, set very low latency at the cost of increased power consumption.

Reviewer 2: ```suggestion
<1> If `highPowerConsumption` is true, the node is tuned for very low latency at the cost of increased power consumption.
```

> <2> Disables certain debugging and monitoring features, but leaves power management on.

Reviewer 2: ```suggestion
<2> Disables certain debugging and monitoring features that can affect system latency.
```

Author: Done.

// Module included in the following assemblies:
//
// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc

:_content-type: CONCEPT
[id="cnf-configuring-workload-hints_{context}"]
= Configuring workload hints

Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:

* Low latency
* Real-time capability
* Efficient use of power

> In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case that workload falls. The Node Tuning Operator uses hints to fine tune the performance settings for the workload.

Reviewer 2: Or did we say The NTO uses PerformanceProfile? Makes little difference to me though.

> In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case that workload falls. The Node Tuning Operator uses hints to fine tune the performance settings for the workload.

Author: Went with: "The Node Tuning Operator uses `PerformanceProfile` to fine tune the performance settings for the workload."

The environment in which an application is operating influences its behavior. For a typical data center with no strict latency requirements, only minimal default tuning is needed that enables CPU partitioning for some high performance workload pods.
For data centers and workloads where latency is a higher priority, measures are still taken to optimize power consumption.
The most complicated cases are clusters close to latency-sensitive equipment such as manufacturing machinery and software-defined radios. This last class of deployment is often referred to as Far edge. For Far edge deployments, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.

[NOTE]
====
The following workload hints can be configured manually or created using the Performance Profile Creator. See "Creating a performance profile" for further information.
====

[cols="1,1,1,1",options="header"]
|===

>     | `PerformanceProfile` setting| Hint | Environment | Description

Reviewer 2: ```suggestion
    | Performance Profile creator setting | Hint | Environment | Description
```

Author: Done.

    | Default
    a|[source,terminal]
----
workloadHints:
highPowerConsumption: false
realtime: false
----
    | High throughput cluster without latency requirements
    | Performance achieved through CPU partitioning only.

    | Low-latency
    a|[source,terminal]
----
workloadHints:
highPowerConsumption: false
realtime: true
----
    | Regional datacenters
    | Both energy savings and low-latency are desirable: compromise between power management, latency and throughput.

    | Ultra-low-latency
    a|[source,terminal]
----
workloadHints:
highPowerConsumption: true
realtime: true
----
    | Far edge clusters, latency critical workloads
    | Optimized for absolute minimal latency and maximum determinism at the cost of increased power consumption.

|===

.Procedure

. Create a `PerformanceProfile` appropriate for the environment's hardware and topology as described in previous sections. Then adjust the profile to match the expected workload. In this example, we tune for the lowest possible latency.

> . Add the `highPowerConsumption` and `realtime` workload hints and set them both to true as indicated by the above table.

Reviewer 3: in the .md is realTime:
`| realTime | RealTime defines if the node should be configured for the real time workload. | *bool | false |`

but in the yaml is realtime. Not sure why highPowerConsumption is camelCase and realtime is not.

Author: If realtime in the yaml I will leave as is.

+
[source,yaml]
----
    apiVersion: performance.openshift.io/v2
    kind: PerformanceProfile
    metadata:
      name: workload-hints
    spec:
      ...
      workloadHints:
        highPowerConsumption: true <1>
        realtime: true <2>
----
<1> If `highPowerConsumption` is true, the node is tuned for very low latency at the cost of increased power consumption.

> <2> Disables some debugging and monitoring features that can affect system latency.

Reviewer 4: I think this module needs to be restructured or broken up into multiple modules. The module begins as a concept. Then, it provides a table of reference materials. Finally, it has a procedure. 

I would also add an ==Additional resources section after the procedure with a link to "Creating a performance profile," because the first step of the procedure says "as described in previous sections", and it isn't in a previous section. It's actually in the subsequent assembly.

Let me know if you need help with this.

Author: As mentioned above, I relocated the conceptual material to the top level conceptual topic on low latency ("Understanding low latency"). I gave the table a topic of its own - it needs to precede the procedure, as you have to use the table to know the right workload hint settings to go for.
